### PR TITLE
Cross-version API Compatibility tests - Feature flags API

### PR DESCRIPTION
**Pull request by mpanchajanya**
### What this PR does / why we need it

- Added Config Features APIs compatibility tests  for SetFeature, IsFeatureEnabled, DeleteFeature APIs for Runtime libraries latest, v0.28.0, v0.25.4, v0.11.6.
- Added helper methods to create API Commands, Validators to validate input and outputOptions.
- Extended the testplugins implementation to trigger Feature APIs.
- Added Default helper functions to create the input and output options data for command creation

Detailed Info:

`test/compatibility`
`core/api_constants.go` -> Added Feature Flag APIs names

`framework` changes
`compatibility-tests/featureflags` -> Add compatibility tests for feature flag APIs

`featureflags` directory to provide functions to create api commands and validations to help write compatibility tests for feature flags APIs

`testplugins` changes
Add Implementation to trigger feature apis for all supported runtime versions

### Which issue(s) this PR fixes

Fixes #

### Describe testing done for PR

- Added Unit testing for the helper functions like Command creations, Trigger Feature APIs.
- Build and Run compatibility tests using below command and make sure all tests pass
`make compatibility-tests`

```
Ginkgo ran 4 suites in 28.183200333s
Test Suite Passed

```

### Release note

```release-note
Added Feature flag Config APIs compatibility tests
```

### CODE REVIEW

File: test/compatibility/framework/featureflags/feature_validators.go [hunk 1]
```diff
@@ -0,0 +1,76 @@
...
+
+package featureflags
+
+import (
+	"fmt"
+
+	"github.com/vmware-tanzu/tanzu-plugin-runtime/test/compatibility/core"
+)
+
+// Validate  the setFeatureInputOptions as per runtime version i.e. check whether mandatory fields are set and throw error if missing
```
Comment (vuil) [inline test/compatibility/framework/featureflags/feature_validators.go:12]: double space

File: test/compatibility/framework/compatibilitytests/featureflags/featureflags_test_helpers.go [hunk 2]
```diff
@@ -0,0 +1,42 @@
...
+}
+
+// DefaultDeleteFeatureCommand creates a DeleteFeature Command with default input and output options
+func DefaultDeleteFeatureCommand(version core.RuntimeVersion, key string) *core.Command {
+	cmd, err := featureflags.NewDeleteFeatureCommand(featureflags.ConstructDeleteFeatureInputOptions(version, common.CompatibilityTestsPlugin, key), nil)
+	gomega.Expect(err).To(gomega.BeNil())
+	return cmd
+}
+
+// DefaultSetFeatureCommand creates a DeleteFeature Command with default input and output options
```
Comment (vuil) [inline test/compatibility/framework/compatibilitytests/featureflags/featureflags_test_helpers.go:37]: DeleteFeature -> SetFeature

File: test/compatibility/framework/compatibilitytests/featureflags/featureflags_test_helpers.go [hunk 3]
```diff
@@ -0,0 +1,42 @@
...
+)
+
+// DefaultIsFeatureEnabledCommand creates a IsFeatureEnabled Command with default input and output options
+func DefaultIsFeatureEnabledCommand(version core.RuntimeVersion, key string, expectedFeatureEnabled bool) *core.Command {
+	cmd, err := featureflags.NewIsFeatureEnabledCommand(featureflags.ConstructIsFeatureEnabledInputOptions(version, common.CompatibilityTestsPlugin, key), featureflags.ConstructIsFeatureEnabledOutputOptions(version, expectedFeatureEnabled))
+	gomega.Expect(err).To(gomega.BeNil())
+	return cmd
+}
+
+// DefaultIsFeatureEnabledCommandWithOutputError creates a DefaultIsFeatureEnabledCommandWithOutputError Command with default input options and output with error options
```
Comment (vuil) [inline test/compatibility/framework/compatibilitytests/featureflags/featureflags_test_helpers.go:23]: DefaultIsFeatureEnabledCommandWithOutputError -> IsFeatureEnabled

File: test/compatibility/framework/featureflags/feature_commands.go [hunk 4]
```diff
@@ -0,0 +1,175 @@
...
+	"strconv"
+
+	"github.com/vmware-tanzu/tanzu-plugin-runtime/test/compatibility/core"
+)
+
+// NewSetFeatureCommand constructs a command to make a call to specific runtime version SetFeature API
+// Input Parameter inputOpts has all input parameters which are required for Runtime SetFeature API
+// Input Parameter: outputOpts has details about expected output from Runtime SetFeature API call
+// Return: command to execute or error if any validations fails for SetFeatureInputOptions or SetFeatureOutputOptions
+// This method does validate the input parameters  SetFeatureInputOptions or SetFeatureOutputOptions based on Runtime API Version
```
Comment (vuil) [inline test/compatibility/framework/featureflags/feature_commands.go:18]: double space

File: test/compatibility/framework/featureflags/feature_commands.go [hunk 5]
```diff
@@ -0,0 +1,175 @@
...
+	c.APIs = append(c.APIs, api)
+
+	return c, nil
+}
+
+// NewIsFeatureEnabledCommand constructs a command to make a call to specific runtime version IsFeatureEnabled API
+// Input Parameter inputOpts has all input parameters which are required for Runtime IsFeatureEnabled API
+// Input Parameter: outputOpts has details about expected output from Runtime IsFeatureEnabled API call
+// Return: command to execute or error if any validations fails for IsFeatureEnabledInputOptions or IsFeatureEnabledOutputOptions
+// This method does validate the input parameters  IsFeatureEnabledInputOptions or IsFeatureEnabledOutputOptions based on Runtime API Version
```
Comment (vuil) [inline test/compatibility/framework/featureflags/feature_commands.go:69]: double space

File: test/compatibility/framework/featureflags/feature_commands.go [hunk 6]
```diff
@@ -0,0 +1,175 @@
...
+
+	c.APIs = append(c.APIs, api)
+	return c, nil
+}
+
+// NewDeleteFeatureCommand constructs a command to make a call to specific runtime version DeleteFeature API
+// Input Parameter inputOpts has all input parameters which are required for Runtime DeleteFeature API
+// Input Parameter: outputOpts has details about expected output from Runtime DeleteFeature API call
+// Return: command to execute or error if any validations fails for DeleteFeatureInputOptions or DeleteFeatureOutputOptions
+// This method does validate the input parameters  DeleteFeatureInputOptions or DeleteFeatureOutputOptions based on Runtime API Version
```
Comment (vuil) [inline test/compatibility/framework/featureflags/feature_commands.go:132]: double space

File: test/compatibility/framework/featureflags/feature_helpers.go [hunk 7]
```diff
@@ -0,0 +1,62 @@
+// Copyright 2023 VMware, Inc. All Rights Reserved.
+// SPDX-License-Identifier: Apache-2.0
+
+package featureflags
+
+import (
+	"github.com/vmware-tanzu/tanzu-plugin-runtime/test/compatibility/core"
+)
+
+// ConstructIsFeatureEnabledInputOptions creates a DeleteFeature Command with default input and output options
```
Comment (vuil) [inline test/compatibility/framework/featureflags/feature_helpers.go:10]: DeleteFeature -> IsFeatureEnabledInputOptions

**Review by vuil — COMMENTED** [review]
Bunch of nits on typos, but otherwise lgtm, thanks!

File: test/compatibility/framework/featureflags/feature_validators.go (hunk 1 above)
Comment (mpanchajanya) [reply test/compatibility/framework/featureflags/feature_validators.go:12]: removed the double space

File: test/compatibility/framework/compatibilitytests/featureflags/featureflags_test_helpers.go (hunk 2 above)
Comment (mpanchajanya) [reply test/compatibility/framework/compatibilitytests/featureflags/featureflags_test_helpers.go:37]: updated the comment

File: test/compatibility/framework/featureflags/feature_commands.go (hunk 4 above)
Comment (mpanchajanya) [reply test/compatibility/framework/featureflags/feature_commands.go:18]: updated

File: test/compatibility/framework/featureflags/feature_commands.go (hunk 5 above)
Comment (mpanchajanya) [reply test/compatibility/framework/featureflags/feature_commands.go:69]: updated

File: test/compatibility/framework/featureflags/feature_commands.go (hunk 6 above)
Comment (mpanchajanya) [reply test/compatibility/framework/featureflags/feature_commands.go:132]: updated

File: test/compatibility/framework/featureflags/feature_helpers.go (hunk 7 above)
Comment (mpanchajanya) [reply test/compatibility/framework/featureflags/feature_helpers.go:10]: updated

File: test/compatibility/framework/compatibilitytests/featureflags/featureflags_test_helpers.go (hunk 3 above)
Comment (mpanchajanya) [reply test/compatibility/framework/compatibilitytests/featureflags/featureflags_test_helpers.go]: Refactored the methods

**Comment by mpanchajanya** [comment]
> Bunch of nits on typos, but otherwise lgtm, thanks!

@vuil Thanks for the review, Addressed the comments

**Review by vuil — APPROVED** [review]
lgtm, thanks.

**Pull request merged**